lib/Parse/ParseStmt.cpp

```diff
@@ -2350,6 +2351,43 @@ parseStmtCaseDefault(Parser &P, SourceLoc &CaseLoc,
   return Status;
 }
 
+namespace {
+
+struct FallthroughFinder : ASTWalker {
```

Comment:
To double check, previous code from SILGen seemed to only need to know if a `fallthrough` existed at all, it did not use the `FallthroughStmt` pointer, is it really necessary to store that pointer in the `CastStmt` ?

Reply:
You are right about the previous code in SILGen.

The reason why I am doing this is so that you guys can implement the replace-all thing. Btw, I discovered that we may also want to be be able to go the other direction as well. This would make it much more efficient to find the case stmt of any of the var decls. It also would just be useful information to have.

Reply:
What I am saying is that previously we spoke about having fallthrough stmt pointers from a fallthrough src -> a fallthrough dst. That would create the cycle with the VarDecls so you could get all of the VarDecls for renaming purposes.

In the process of implementing more of what we spoke about, I discovered that getting the CaseStmt in the manner in which we talked about would be... unfortunate. We would have to go through all of the parent VarDecls until we found the one with the CaseStmt parent. Then we would have to go down the CaseStmts, comparing the VarDecls of the pattern in each CaseLabelItem of each CaseStmt until we found it. That just seems really inefficient. If we instead just had it so that:

1. The first case label item's var decls all point at the CaseStmt.
2. We make the CaseStmt fallthrough chain be a double linked list (which if you are worried about memory, I should be able to compress them.

Then we can still visit the entire cycle. The only difference is that when we hit a var decl whose parent is the case stmt, if that parent is a fallthrough destination, we map the given var decl to its corresponding CaseBodyVarDecl. We can do this with some helpers.

Your thoughts?

Reply:
To be clear, the reason why we can map the var decl from the fallthrough destination case stmt to a CaseBodyVarDecl in the fallthrough source case stmt is b/c we know that for each set of {VarDecls} each name is unique and it is assumed that the VarDecls with the same names map to each other. This is guaranteed by the type checker. Thus we can always find the appropriate index.

Reply:
Argyrios and I spoke offline. He agreed with me that we actually need the fallthrough pointer. I agreed with him that my performance concern is not that important here since we are not likely to hit the case of having multiple fallthroughs at all meaning if we have to go all the way to the first fallthrough and then back down to find a parent pattern, it is not an issue.